service-userlayer/src/main/java/org/oskari/map/userlayer/input/KMLParser.java

```diff
@@ -28,24 +40,126 @@
+            // Containers (Document, Folder) can contain Features
+            // --> have to loop Features and check the type and if Feature is container then loop it's features...
+
+            // Using PullParser to parse Placemarks only
+            PullParser parser = new PullParser(new KMLConfiguration(), in, KML.Placemark);
+            SimpleFeature f;
+            Set<String> extendedData = new HashSet<String>();
+            Map<String, Object> untypedData;
+
             DefaultFeatureCollection fc = new DefaultFeatureCollection();
-            SimpleFeature f = (SimpleFeature) parser.parse(in);
-            fc.add(f);
-            return fc;
+            while ((f = (SimpleFeature) parser.parse()) != null) {
+                if (f.getUserData().get("UntypedExtendedData") != null) {
+                    untypedData = (Map<String, Object>) f.getUserData().get("UntypedExtendedData");
+                    untypedData.keySet().forEach((k) -> extendedData.add(k));
+                }
+                fc.add(f);
+            }
+            return processFeatures(targetCRS, fc, extendedData);
+        } catch (XMLStreamException e) {
+            throw new ServiceException("XMLStreamException occured", e);
         } catch (IOException e) {
             throw new ServiceException("IOException occured", e);
         } catch (SAXException e) {
-            LOG.warn(e, "Failed to parse KML file");
             throw new ServiceException("Invalid KML file: " + e.getMessage());
-        } catch (ParserConfigurationException e) {
-            throw new ServiceException("KML parser failed to initialize");
         }
     }
 
+    @SuppressWarnings("unchecked")
+    private DefaultFeatureCollection processFeatures (CoordinateReferenceSystem targetCRS, DefaultFeatureCollection fc_kml, Set<String> extendedData) throws ServiceException {
+        try {
+            DefaultFeatureCollection fc = new DefaultFeatureCollection();
+            SimpleFeatureBuilder builder;
+            SimpleFeature f;
+            // KML always lon,lat 4326
+            CoordinateReferenceSystem sourceCRS = CRS.decode("EPSG:4326", true);
+            MathTransform transform = FeatureCollectionParsers.getTransform(sourceCRS, targetCRS);
+
+            if (extendedData.isEmpty()){
+               builder = getBasicBuilder(targetCRS);
+            } else {
+               builder = getExtendedBuilder(targetCRS, extendedData);
+            }
+            SimpleFeatureIterator iter = fc_kml.features();
+            while (iter.hasNext()){
+                f = iter.next();
+                builder.set(KML_NAME, f.getAttribute("name"));
+                // Basic builder
+                if (extendedData.isEmpty()){
+                    builder.set(KML_DESC, f.getAttribute("description"));
+                // Extended builder
+                } else {
+                    Map<String, Object> untypedData = (Map<String, Object>) f.getUserData().get("UntypedExtendedData");
```

Comment:
This is a bit difficult to follow. Maybe add a comment explaining _why_ this is done?

Reply:
Looks like it just tries to dig up a description from multiple places? Does calling set(KML_DESC, [something]) several times trigger an exception? If not this could be simplified by always doing:

    builder.set(KML_NAME, f.getAttribute("name")); 
    builder.set(KML_DESC, f.getAttribute("description"));

and overriding the desc value if there's a description found on the extended data? It's a bit confusing that untypedData is checked for non-null -> extendedData is looped, but the value is fetched from untypedData with the key from extendedData. Are the keys in untypedData not in sync with extendedData? Maybe refactor this logic to a new method with some comment on it.